### ISSUE TITLE
How to specify table name format in base model depending on child model's overriden table name ?

### ISSUE THREAD

**Issue by avinashs2401** [jazzband/django-simple-history#1174]
Hi,
I have a Base model like this, which many models would inherit from,

```
class BaseModel(models.Model):
    created_date = models.DateTimeField(auto_now_add=True)
    last_updated_date = models.DateTimeField(auto_now=True)
    history = HistoricalRecords(inherit=True)
```

I have child models inheriting from this BaseModel and I have overriden the db table names to be with underscores. Eg.
```
class ProductVariants(BaseModel):
   title = models.CharField(max_length=255, blank=False, null=False)
   class Meta:
       db_table = "product_variants"
```

I want History table to be created as `appname_historical_product_variants`. But right now its getting created as `appname_historicalproductvariants` 

Any suggestions on how to achieve this?

**Comment by singhravi1** [comment]
@avinashs2401   
You can override `HistoricalRecords` class and add an attribute,  
```
DEFAULT_MODEL_NAME_PREFIX = "Historical_"
```

Now use this new class in your models for the history field.

**Comment by singhravi1** [comment]
@avinashs2401 
You can use `HistoricalRecords` attribute `table_name` to mention the table name for the historical model.

**Comment by singhravi1** [comment]
@avinashs2401 
If you manage it in the base model only, you can override the `HistoricalRecords` class and specify a custom attribute `DEFAULT_MODEL_NAME_PREFIX`.
Otherwise you can mention the `history` field specifically each model with `table_name` attribute.

**Comment by muneeb706** [comment]
@avinashs2401 
As per my understanding, there is no way to specify custom table name for history models of inherited child models.
You can only change the history table name where HistoricalRecords instance is created (i.e. in the Base Class).

Overriding db_table property of Meta will only update table name for the model itself ( this is done by django: https://docs.djangoproject.com/en/4.2/ref/models/options/#db-table).

**Comment by singhravi1** [comment]
@muneeb706 
You could override the `HistoricalRecords`'s `create_history_model` method and add a logic using the model name being generated, to change the `attrs["Meta"].db_table` as per your requirements

**Comment by muneeb706** [comment]
@singhravi1 

Do you mean, something like this ?

```
from django.db import models
from simple_history.models import HistoricalRecords


class CustomHistoricalRecords(HistoricalRecords):
    def create_history_model(self, model, inherited):
        history_model = super().create_history_model(model, inherited)
        history_model._meta.db_table = "historical_" + model._meta.db_table
        return history_model

class BaseModel(models.Model):
    created_date = models.DateTimeField(auto_now_add=True)
    last_updated_date = models.DateTimeField(auto_now=True)
    history = CustomHistoricalRecords(inherit=True)
  
class ChildModel(BaseModel):
    title = models.CharField(max_length=255, blank=False, null=False)
    
    class Meta:
       db_table = "child_model"
```

**Comment by muneeb706** [comment]
> @singhravi1
> 
> Do you mean, something like this ?
> 
> ```
> from django.db import models
> from simple_history.models import HistoricalRecords
> 
> 
> class CustomHistoricalRecords(HistoricalRecords):
>     def create_history_model(self, model, inherited):
>         history_model = super().create_history_model(model, inherited)
>         history_model._meta.db_table = "historical_" + model._meta.db_table
>         return history_model
> 
> class BaseModel(models.Model):
>     created_date = models.DateTimeField(auto_now_add=True)
>     last_updated_date = models.DateTimeField(auto_now=True)
>     history = CustomHistoricalRecords(inherit=True)
>   
> class ChildModel(BaseModel):
>     title = models.CharField(max_length=255, blank=False, null=False)
>     
>     class Meta:
>        db_table = "child_model"
> ```

Please, note, the meta class db_table attribute will not be updated.

**Comment by muneeb706** [comment]
> > @singhravi1
> > Do you mean, something like this ?
> > ```
> > from django.db import models
> > from simple_history.models import HistoricalRecords
> > 
> > 
> > class CustomHistoricalRecords(HistoricalRecords):
> >     def create_history_model(self, model, inherited):
> >         history_model = super().create_history_model(model, inherited)
> >         history_model._meta.db_table = "historical_" + model._meta.db_table
> >         return history_model
> > 
> > class BaseModel(models.Model):
> >     created_date = models.DateTimeField(auto_now_add=True)
> >     last_updated_date = models.DateTimeField(auto_now=True)
> >     history = CustomHistoricalRecords(inherit=True)
> >   
> > class ChildModel(BaseModel):
> >     title = models.CharField(max_length=255, blank=False, null=False)
> >     
> >     class Meta:
> >        db_table = "child_model"
> > ```
> 
> Please, note, the meta class db_table attribute will not be updated.

I don't think we can access attrs attribute as it is defined inside the create_history_model method.

**Comment by muneeb706** [comment]
@singhravi1 
If you update `self.table_name` then the db_table will be updated. However, in your code snippet, you have added the condition. 
`if not self.table_name`  which will be True only when the table_name is not set in the base model HistoricalRecords instance.

However, you can make it work by updating `self.table_name` property as in `HistoricalRecords.create_history_model` method
`attrs["Meta"].db_table = self.table_name`. This may not be the best way to do it, as according to current implementation self.table_name is supposed to point to value  that is set when HistoricalRecords instance is created (which is in base class in this scenario).

**Comment by singhravi1** [comment]
Hey @muneeb706 ,
The check can be changed based on the logic that mentioning `table_name` with `inherit=True` would not make sense.
Not the best way but till the time this feature is included in the package, users such as the author can extend it to have what they require.
The following updated code should work for the problem in hand,
```
class CustomHistoricalRecords(HistoricalRecords):
    def create_history_model(self, model, inherited):
        if self.inherit:
            self.table_name = f"{model._meta.app_label}_{re.sub(r'(?<!^)(?=[A-Z])', '_', self.get_history_model_name(model)).lower()}"
        history_model = super().create_history_model(model, inherited)
        if not self.inherit:
            self.table_name = None
        return history_model
```
